// v-money3/dist/v-money3.mjs
import { computed as e, createElementBlock as t, defineComponent as n, getCurrentInstance as r, mergeProps as i, onMounted as a, openBlock as o, ref as s, resolveDirective as c, toRefs as l, unref as u, useAttrs as d, watch as f, withDirectives as p } from "vue";
//#region src/options.ts
var m = {
	debug: !1,
	masked: !1,
	prefix: "",
	suffix: "",
	thousands: ",",
	decimal: ".",
	precision: 2,
	disableNegative: !1,
	disabled: !1,
	min: null,
	max: null,
	setMaxIfBigger: !0,
	allowBlank: !1,
	treatZeroAsBlank: !0,
	minimumNumberOfCharacters: 0,
	modelModifiers: { number: !1 },
	shouldRound: !0,
	focusOnRight: !1,
	lazy: !0
}, h = ["+", "-"], g = [
	"decimal",
	"thousands",
	"prefix",
	"suffix"
];
function _(e) {
	return Math.max(0, Math.min(e, 100));
}
function v(e, t) {
	return e = e.padStart(t + 1, "0"), t === 0 ? e : `${e.slice(0, -t)}.${e.slice(-t)}`;
}
function y(e) {
	return e = e ? e.toString() : "", e.replace(/\D+/g, "") || "0";
}
function b(e, t) {
	return e.replace(/(\d)(?=(?:\d{3})+\b)/gm, `$1${t}`);
}
function x(e, t, n) {
	return t ? e + n + t : e;
}
function S(e, t) {
	return h.includes(e) ? (console.warn(`v-money3 "${t}" property don't accept "${e}" as a value.`), !1) : /\d/g.test(e) ? (console.warn(`v-money3 "${t}" property don't accept "${e}" (any number) as a value.`), !1) : !0;
}
function C(e) {
	for (let t of g) if (!S(e[t], t)) return !1;
	return !0;
}
function w(e) {
	for (let t of g) {
		if (typeof e[t] != "string") {
			e[t] = "";
			continue;
		}
		e[t] = e[t].replace(/\d+/g, "");
		for (let n of h) e[t] = e[t].replaceAll(n, "");
	}
	return e;
}
function T(e) {
	return e.length - (e.indexOf(".") + 1);
}
function E(e) {
	return e.replace(/^(-?)0+(?!\.)(.+)/, "$1$2");
}
function D(e) {
	return /^-?[\d]+$/g.test(e);
}
function O(e) {
	return /^-?[\d]+(\.[\d]+)$/g.test(e);
}
function k(e, t, n) {
	return t > e.length - 1 ? e : e.substring(0, t) + n + e.substring(t + 1);
}
function A(e, t) {
	let n = t - T(e);
	if (n >= 0) return e;
	let r = e.slice(0, n), i = e.slice(n);
	if (r.charAt(r.length - 1) === "." && (r = r.slice(0, -1)), parseInt(i.charAt(0), 10) >= 5) {
		for (let e = r.length - 1; e >= 0; --e) {
			let t = r.charAt(e);
			if (t !== "." && t !== "-") {
				let n = parseInt(t, 10) + 1;
				if (n < 10) return k(r, e, n);
				r = k(r, e, "0");
			}
		}
		return `1${r}`;
	}
	return r;
}
function j(e, t) {
	let n = () => {
		e === document.activeElement && e.setSelectionRange(t, t);
	};
	e === document.activeElement && (n(), setTimeout(n, 1));
}
function M(e) {
	return new Event(e, {
		bubbles: !0,
		cancelable: !1
	});
}
function N({ debug: e = !1 }, ...t) {
	e && console.log(...t);
}
//#endregion
//#region \0@oxc-project+runtime@0.130.0/helpers/typeof.js
function P(e) {
	"@babel/helpers - typeof";
	return P = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(e) {
		return typeof e;
	} : function(e) {
		return e && typeof Symbol == "function" && e.constructor === Symbol && e !== Symbol.prototype ? "symbol" : typeof e;
	}, P(e);
}
//#endregion
//#region \0@oxc-project+runtime@0.130.0/helpers/toPrimitive.js
function F(e, t) {
	if (P(e) != "object" || !e) return e;
	var n = e[Symbol.toPrimitive];
	if (n !== void 0) {
		var r = n.call(e, t || "default");
		if (P(r) != "object") return r;
		throw TypeError("@@toPrimitive must return a primitive value.");
	}
	return (t === "string" ? String : Number)(e);
}
//#endregion
//#region \0@oxc-project+runtime@0.130.0/helpers/toPropertyKey.js
function I(e) {
	var t = F(e, "string");
	return P(t) == "symbol" ? t : t + "";
}
//#endregion
//#region \0@oxc-project+runtime@0.130.0/helpers/defineProperty.js
function L(e, t, n) {
	return (t = I(t)) in e ? Object.defineProperty(e, t, {
		value: n,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[t] = n, e;
}
//#endregion
//#region src/BigNumber.ts
var R = class e {
	constructor(e) {
		L(this, "number", 0n), L(this, "decimal", 0), this.setNumber(e);
	}
	getNumber() {
		return this.number;
	}
	getDecimalPrecision() {
		return this.decimal;
	}
	setNumber(e) {
		this.decimal = 0, typeof e == "bigint" ? this.number = e : typeof e == "number" ? this.setupString(e.toString()) : this.setupString(e);
	}
	toFixed(e = 0, t = !0) {
		let n = this.toString(), r = e - this.getDecimalPrecision();
		if (r > 0) return n.includes(".") || (n += "."), n.padEnd(n.length + r, "0");
		if (r < 0) {
			if (t) return A(n, e);
			let i = n.slice(0, r);
			return i.endsWith(".") ? i.slice(0, -1) : i;
		}
		return n;
	}
	toString() {
		let e = this.number.toString();
		if (this.decimal) {
			let t = !1;
			return e.charAt(0) === "-" && (e = e.substring(1), t = !0), e = e.padStart(e.length + this.decimal, "0"), e = `${e.slice(0, -this.decimal)}.${e.slice(-this.decimal)}`, e = E(e), (t ? "-" : "") + e;
		}
		return e;
	}
	lessThan(e) {
		let [t, n] = this.adjustComparisonNumbers(e);
		return t < n;
	}
	biggerThan(e) {
		let [t, n] = this.adjustComparisonNumbers(e);
		return t > n;
	}
	isEqual(e) {
		let [t, n] = this.adjustComparisonNumbers(e);
		return t === n;
	}
	setupString(e) {
		if (e = E(e), D(e)) this.number = BigInt(e);
		else if (O(e)) this.decimal = T(e), this.number = BigInt(e.replace(".", ""));
		else throw Error(`BigNumber has received an invalid format for the constructor: ${e}`);
	}
	adjustComparisonNumbers(t) {
		let n;
		n = t instanceof e ? t : new e(t);
		let r = this.getDecimalPrecision() - n.getDecimalPrecision(), i = this.getNumber(), a = n.getNumber();
		return r > 0 ? a = n.getNumber() * 10n ** BigInt(r) : r < 0 && (i = this.getNumber() * 10n ** BigInt(r * -1)), [i, a];
	}
};
//#endregion
//#region src/format.ts
function z(e, t = m, n = "") {
	N(t, "utils format() - caller", n), N(t, "utils format() - input1", e);
	let r = _(t.precision);
	if ((e == null || e === "") && t.allowBlank) return "";
	if (e == null) e = "";
	else if (typeof e == "number") e = t.shouldRound ? e.toFixed(r) : e.toFixed(Math.min(r + 1, 100)).slice(0, -1);
	else if (t.modelModifiers && t.modelModifiers.number && D(e)) e = Number(e).toFixed(r);
	else if (!t.disableNegative && e === "-") return e;
	N(t, "utils format() - input2", e);
	let i = t.disableNegative ? "" : e.indexOf("-") >= 0 ? "-" : "", a = e.replace(t.prefix, "").replace(t.suffix, "");
	N(t, "utils format() - filtered", a), !r && t.thousands !== "." && O(a) && (a = A(a, 0), N(t, "utils format() - !precision && isValidFloat()", a));
	let o = y(a);
	N(t, "utils format() - numbers", o), N(t, "utils format() - numbersToCurrency", i + v(o, r));
	let s = new R(i + v(o, r));
	N(t, "utils format() - bigNumber1", s.toString()), t.setMaxIfBigger !== !1 && t.max !== null && t.max !== void 0 && t.max !== "" && s.biggerThan(t.max) && s.setNumber(t.max), t.min !== null && t.min !== void 0 && t.min !== "" && s.lessThan(t.min) && s.setNumber(t.min), t.disableNegative && s.lessThan(0) && s.setNumber(0);
	let c = s.toFixed(r, t.shouldRound);
	if (N(t, "utils format() - bigNumber2", s.toFixed(r)), /^0(\.0+)?$/g.test(c) && t.allowBlank && t.treatZeroAsBlank) return "";
	let [l, u] = c.split("."), d = u === void 0 ? 0 : u.length, f = l.charAt(0) === "-", p = (f ? l.slice(1) : l).padStart(t.minimumNumberOfCharacters - d, "0");
	l = (f ? "-" : "") + b(p, t.thousands);
	let h = t.prefix + x(l, u, t.decimal) + t.suffix;
	return N(t, "utils format() - output", h), h;
}
//#endregion
//#region src/unformat.ts
function B(e, t = m, n = "") {
	if (N(t, "utils unformat() - caller", n), N(t, "utils unformat() - input", e), !t.disableNegative && e === "-") return N(t, "utils unformat() - return netagive symbol", e), e;
	let r = t.disableNegative ? "" : e.indexOf("-") >= 0 ? "-" : "", i = e.replace(t.prefix, "").replace(t.suffix, "");
	N(t, "utils unformat() - filtered", i);
	let a = y(i);
	N(t, "utils unformat() - numbers", a);
	let o = new R(r + v(a, t.precision));
	N(t, "utils unformat() - bigNumber1", a.toString()), t.setMaxIfBigger !== !1 && t.max !== null && t.max !== void 0 && t.max !== "" && o.biggerThan(t.max) && o.setNumber(t.max), t.min !== null && t.min !== void 0 && t.min !== "" && o.lessThan(t.min) && o.setNumber(t.min), t.disableNegative && o.lessThan(0) && o.setNumber(0);
	let s = o.toFixed(_(t.precision), t.shouldRound);
	return t.modelModifiers && t.modelModifiers.number && (s = parseFloat(s)), N(t, "utils unformat() - output", s), s;
}
//#endregion
//#region src/directive.ts
var V = [
	"precision",
	"decimal",
	"thousands",
	"prefix",
	"suffix",
	"min",
	"max",
	"setMaxIfBigger",
	"allowBlank",
	"treatZeroAsBlank",
	"minimumNumberOfCharacters",
	"shouldRound",
	"modelModifiers"
], H = "__v_money3_last_valid__", U = "__v_money3_is_wrapper__", W = "__v_money3_synth__", G = (e, t, n) => {
	if (N(t, "directive setValue() - caller", n), !C(t)) {
		N(t, "directive setValue() - validateRestrictedOptions() return false. Stopping here...", e.value);
		return;
	}
	let r = e.value.length - (e.selectionEnd || 0), i = z(e.value, t, n);
	if (t.setMaxIfBigger === !1 && t.max !== null && t.max !== void 0 && t.max !== "") {
		let n = B(i, t, "directive setValue overflow check");
		if (new R(String(n)).biggerThan(t.max)) {
			let t = e[H];
			typeof t == "string" && t !== e.value && (e.value = t);
			return;
		}
	}
	if (i === e.value) {
		let t = e[H];
		e[H] = i, t !== void 0 && t !== i && e.dispatchEvent(M("change"));
		return;
	}
	if (e.value = i, e[H] = i, r = Math.max(r, t.suffix.length), r = e.value.length - r, r = Math.max(r, t.prefix.length), j(e, r), e.dispatchEvent(M("change")), e[U]) {
		let t = new Event("input", { bubbles: !1 });
		t[W] = !0, e.dispatchEvent(t);
	}
}, K = (e, t) => {
	let n = e.currentTarget, r = e.code === "Backspace" || e.code === "Delete", i = n.value.length - (n.selectionEnd || 0) === 0;
	N(t, "directive onkeydown() - el.value", n.value), N(t, "directive onkeydown() - backspacePressed", r), N(t, "directive onkeydown() - isAtEndPosition", i), t.allowBlank && t.treatZeroAsBlank && r && i && parseFloat(String(B(n.value, t, "directive onkeydown allowBlank"))) === 0 && (N(t, "directive onkeydown() - set el.value = \"\"", n.value), n.value = "", n.dispatchEvent(M("change"))), N(t, "directive onkeydown() - e.key", e.key), e.key === "+" && n.value.indexOf("-") >= 0 && (N(t, "directive onkeydown() - flipping sign on el.value", n.value), n.value = n.value.replace("-", ""), G(n, t, "directive onkeydown +"));
}, q = (e, t) => {
	if (e[W]) return;
	let n = e.currentTarget;
	N(t, "directive oninput()", n.value), /^[1-9]$/.test(n.value) && (n.value = v(n.value, _(t.precision)), N(t, "directive oninput() - is 1-9", n.value)), G(n, t, "directive oninput");
}, J = (e, t) => {
	let n = e.currentTarget;
	N(t, "directive onFocus()", n.value), t.focusOnRight && j(n, n.value.length - t.suffix.length);
}, Y = (e) => {
	if (e.tagName.toLocaleUpperCase() !== "INPUT") {
		let t = e.getElementsByTagName("input");
		if (t.length !== 1) throw Error(`v-money3 requires 1 input, found ${t.length} elements.`);
		return t[0];
	}
	return e;
}, ee = (e, t) => {
	e.onkeydown = (e) => {
		K(e, t);
	}, e.oninput = (e) => {
		q(e, t);
	}, e.onfocus = (e) => {
		J(e, t);
	};
};
function te(e, t) {
	return e ? V.some((n) => JSON.stringify(e[n]) !== JSON.stringify(t[n])) : !1;
}
var X = "__v_money3_input__";
function Z(e) {
	let t = e[X];
	if (t) return t;
	let n = Y(e);
	return e[X] = n, n;
}
var Q = {
	mounted(e, t) {
		if (!t.value) return;
		let n = w({
			...m,
			...t.value
		});
		N(n, "directive mounted() - opt", n);
		let r = Z(e);
		r[U] = e !== r, ee(r, n), N(n, "directive mounted() - el.value", r.value), G(r, n, "directive mounted");
	},
	updated(e, t) {
		if (!t.value) return;
		let n = w({
			...m,
			...t.value
		});
		N(n, "directive updated() - opt", n), N(n, "directive updated() - host.value", e.value);
		let r = Z(e);
		if (r[U] = e !== r, z(r.value, n, "directive updated check") !== r.value) {
			if (te(t.oldValue ? w({
				...m,
				...t.oldValue
			}) : null, n) && r.value !== "") {
				console.warn("v-money3: runtime change of format options on the bare directive is unsupported and was skipped to avoid corrupting the value. Re-mount the directive or use the Money3 component instead.");
				return;
			}
			G(r, n, "directive updated");
		}
	},
	beforeUnmount(e) {
		let t = e[X] || e;
		t.onkeydown = null, t.oninput = null, t.onfocus = null, delete t[U], delete e[X];
	}
}, ne = [
	"id",
	"value",
	"disabled"
], $ = /* @__PURE__ */ n({
	inheritAttrs: !1,
	name: "Money3",
	directives: { money3: Q },
	__name: "component",
	props: {
		debug: {
			required: !1,
			type: Boolean,
			default: !1
		},
		id: {
			required: !1,
			type: [Number, String],
			default: () => {
				let e = r();
				return e ? e.uid : null;
			}
		},
		modelValue: {
			required: !0,
			type: [Number, String]
		},
		modelModifiers: {
			required: !1,
			type: Object,
			default: () => ({ number: !1 })
		},
		masked: {
			type: Boolean,
			default: !1
		},
		precision: {
			type: Number,
			default: () => m.precision
		},
		decimal: {
			type: String,
			default: () => m.decimal,
			validator(e) {
				return S(e, "decimal");
			}
		},
		thousands: {
			type: String,
			default: () => m.thousands,
			validator(e) {
				return S(e, "thousands");
			}
		},
		prefix: {
			type: String,
			default: () => m.prefix,
			validator(e) {
				return S(e, "prefix");
			}
		},
		suffix: {
			type: String,
			default: () => m.suffix,
			validator(e) {
				return S(e, "suffix");
			}
		},
		disableNegative: {
			type: Boolean,
			default: !1
		},
		disabled: {
			type: Boolean,
			default: !1
		},
		max: {
			type: [Number, String],
			default: () => m.max
		},
		min: {
			type: [Number, String],
			default: () => m.min
		},
		setMaxIfBigger: {
			type: Boolean,
			default: () => m.setMaxIfBigger
		},
		allowBlank: {
			type: Boolean,
			default: () => m.allowBlank
		},
		treatZeroAsBlank: {
			type: Boolean,
			default: () => m.treatZeroAsBlank
		},
		minimumNumberOfCharacters: {
			type: Number,
			default: () => m.minimumNumberOfCharacters
		},
		shouldRound: {
			type: Boolean,
			default: () => m.shouldRound
		},
		focusOnRight: {
			type: Boolean,
			default: () => m.focusOnRight
		}
	},
	emits: ["update:model-value"],
	setup(n, { emit: r }) {
		let m = n, { modelValue: h, modelModifiers: g, masked: v, precision: y, shouldRound: b, focusOnRight: x } = l(m);
		N(m, "component setup()", m);
		let { value: S } = h;
		(m.disableNegative || S !== "-") && g.value && g.value.number && (S = b.value ? Number(h.value).toFixed(_(y.value)) : Number(h.value).toFixed(Math.min(_(y.value) + 1, 100)).slice(0, -1));
		let C = s(z(S, m, "component setup"));
		N(m, "component setup() - data.formattedValue", C.value);
		function T(e) {
			N(m, "component watch() -> value", e);
			let t = z(e, w({ ...m }), "component watch");
			t !== C.value && (N(m, "component watch() changed -> formatted", t), C.value = t);
		}
		f(h, T);
		let E = null, D = r;
		function O() {
			let e = w({ ...m }), t = h.value;
			if (t === "-") return;
			let n = z(t, e, "component opts watch");
			if (n !== C.value && (C.value = n), v.value && !(g.value && g.value.number)) {
				n !== t && n !== E && (E = n, D("update:model-value", n));
				return;
			}
			let r = B(n, e, "component opts watch reunformat"), i = Number(t), a = Number(r);
			!Number.isNaN(i) && !Number.isNaN(a) && i !== a && a !== E && (E = a, D("update:model-value", r));
		}
		f(() => [
			m.precision,
			m.decimal,
			m.thousands,
			m.prefix,
			m.suffix,
			m.min,
			m.max,
			m.setMaxIfBigger,
			m.allowBlank,
			m.treatZeroAsBlank,
			m.minimumNumberOfCharacters,
			m.shouldRound
		], O), f(() => m.modelModifiers, O, { deep: !0 }), a(() => {
			let e = w({ ...m }), t = B(C.value, e, "component mounted reconcile"), n = Number(h.value), r = Number(t);
			Number.isNaN(n) && !Number.isNaN(r) && r !== E && (E = r, D("update:model-value", t));
		});
		function k(e) {
			let t = e.target.value;
			N(m, "component change() -> evt.target.value", t), v.value && !g.value.number || (t = B(t, w({ ...m }), "component change")), t !== E && (E = t, N(m, "component change() -> update:model-value", t), D("update:model-value", t));
		}
		let A = d(), j = e(() => {
			let e = { ...A };
			return delete e["onUpdate:modelValue"], e;
		});
		return (e, r) => {
			let a = c("money3");
			return p((o(), t("input", i({ id: `${n.id}` }, j.value, {
				type: "tel",
				class: "v-money3",
				value: C.value,
				disabled: m.disabled,
				onChange: k
			}), null, 16, ne)), [[a, {
				precision: u(y),
				decimal: m.decimal,
				thousands: m.thousands,
				prefix: m.prefix,
				suffix: m.suffix,
				disableNegative: m.disableNegative,
				min: m.min,
				max: m.max,
				setMaxIfBigger: m.setMaxIfBigger,
				allowBlank: m.allowBlank,
				treatZeroAsBlank: m.treatZeroAsBlank,
				minimumNumberOfCharacters: m.minimumNumberOfCharacters,
				debug: m.debug,
				modelModifiers: u(g),
				shouldRound: u(b),
				focusOnRight: u(x)
			}]]);
		};
	}
}), re = { install(e) {
	e.component("money3", $), e.directive("money3", Q);
} };
//#endregion
export { R as BigNumber, $ as Money, $ as Money3, $ as Money3Component, Q as Money3Directive, Q as VMoney, Q as VMoney3, re as default, z as format, B as unformat };
